### ISSUE TITLE
ONNX model performance degrades using OpenCV DNN C++ inference

### ISSUE THREAD

**Issue by coordxyz** [ultralytics/yolov5#7548]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I test the pretrained weights of yolov5m6 on MSCOCO val2017. But I get different performance with .pt and .onnx weights. 
Does anyone know why? Thx~
conf-thres=0.25 and iou-thres=0.45 for both of them.
The performance of **.pt** is
![yolov5m6-python](https://user-images.githubusercontent.com/47654539/164911684-eca51d19-56ad-40f1-ae60-809e6aba4ffc.png)
The performance of **C++ with .onnx** is
![yolov5m6-onnxcpp](https://user-images.githubusercontent.com/47654539/164911745-635cf5c8-59ad-4099-82c4-8d448b178de4.png)
p.s. I use cpp inference example from https://github.com/Hexmagic/ONNX-yolov5/blob/master/src/test.cpp.

**Comment by glenn-jocher** [comment]
@coordxyz our official benchmarks validate no accuracy loss on exports on all formats.See https://github.com/ultralytics/yolov5/pull/6613 for details.

### Colab Pro+ V100 High-RAM CPU Results
```
benchmarks: weights=/content/yolov5/yolov5s.pt, imgsz=640, batch_size=1, data=/content/yolov5/data/coco128.yaml, device=cpu, half=False, test=False
Checking setup...
YOLOv5 🚀 v6.1-135-g7926afc torch 1.10.0+cu111 CPU
Setup complete ✅ (8 CPUs, 51.0 GB RAM, 41.5/166.8 GB disk)

Benchmarks complete (241.20s)
                   Format  mAP@0.5:0.95  Inference time (ms)
0                 PyTorch        0.4623               127.61
1             TorchScript        0.4623               131.23
2                    ONNX        0.4623                69.34
3                OpenVINO        0.4623                66.52
4                TensorRT           NaN                  NaN
5                  CoreML           NaN                  NaN
6   TensorFlow SavedModel        0.4623               123.79
7     TensorFlow GraphDef        0.4623               121.57
8         TensorFlow Lite        0.4623               316.61
9     TensorFlow Edge TPU           NaN                  NaN
10          TensorFlow.js           NaN                  NaN
```

**Comment by coordxyz** [comment]
@glenn-jocher Thanks for reply. Did you test the exported onnx model in **C++**? My experiments are using OpenCV DNN **C++** inference. But I can't get the same performance with pytorch model. Do you known why? Thx~
I compare the super parameters of conf_thres, iou_thres, multi-label and w/o box offset in NMS. More experimental results are shown below.
exp1. 
![image](https://user-images.githubusercontent.com/47654539/165434055-d4555585-1d55-482f-b345-5e0085dcde1e.png)
exp2.
![image](https://user-images.githubusercontent.com/47654539/165434142-01d126bd-0a15-4a68-a249-d299c0f5bcc6.png)
exp3.
![image](https://user-images.githubusercontent.com/47654539/165434201-96ca49d8-d4eb-4c69-89d4-b644fb9baeef.png)
exp4. After adding box offset in NMS, C++ performance increases 0.012(mAP) and 0,016(mAR) but is still lower than Python's.
![image](https://user-images.githubusercontent.com/47654539/165434230-592b3931-5772-43b5-a2bb-78d8f33c289b.png)

**Comment by glenn-jocher** [comment]
@coordxyz no, we have not validated in C++, but we have validated OpenCV DNN in Python with i.e. `python val.py --weights yolov5s.onnx --dnn` 

What is boxoffset?

**Comment by coordxyz** [comment]
@glenn-jocher Thx~
The 'boxoffset' is in utils.general.non_max_suppression() as shown in below.
c = x[:, 5:6] * (0 if agnostic else max_wh)    # classes, max_wh=7680
boxes, scores = x[:, :4] + c, x[:, 4]                # boxes (**offset by class**), scores

**Comment by glenn-jocher** [comment]
@coordxyz oh! Box offset is requirement for NMS with more than 1 class. Otherwise there will be drops in mAP due to implementation errors, i.e. it's not supposed to be removable without causing massive mAP drops. Are you sure your results are correct?

**Comment by coordxyz** [comment]
@glenn-jocher I check the bbox visualization and confirm that my results are correct. My C++ code finally achieves the same mAP(0.458) with Python by using opencv Rect2d (double type) rather than Rect (int type) for bbox. But the mAR still drops. The results are shown below.
![image](https://user-images.githubusercontent.com/47654539/165688854-04574745-88f3-4ff8-ae10-b7d25cbc02f1.png)

**Comment by captain0305** [comment]
> @glenn-jocher I check the bbox visualization and confirm that my results are correct. My C++ code finally achieves the same mAP(0.458) with Python by using opencv Rect2d (double type) rather than Rect (int type) for bbox. But the mAR still drops. The results are shown below.
> ![image](https://user-images.githubusercontent.com/47654539/165688854-04574745-88f3-4ff8-ae10-b7d25cbc02f1.png)

I'm interested in the box offset code in cpp, could u please share it with me?

**Comment by glenn-jocher** [comment]
@captain0305 great news! So the error was due to numerical precision, which makes sense. 

I don't know what box offset code you're talking about though, we don't have any C++ in the master YOLOv5 repo.

**Comment by xznhj8129** [comment]
I also get this problem, but going back to pure python i find that there is a big difference when i use dnn versus not; i don't have precise data but i can plainly see that if i run detect.py on the model i trained, as a .pt file, and as .onnx without --dnn it works fine, however with --dnn, using OpenCV's DNN, it's labels are jumbled and it identifies cars as motorcycles, sinks or cakes; in my C++ it will identify cars as toasters... I'm quite a beginner so i'm not sure what could be going on here

**Comment by glenn-jocher** [comment]
@xznhj8129 it sounds like you're experiencing label misidentification when using OpenCV's DNN in both Python and C++. 

This could be caused by input data being in an unexpected format or by misalignment between the model's output and the expected labels. 

Double check that your input data preprocessing and postprocessing procedures match the ones used during training, and that the classes are in the expected order for both the model outputs and your label mapping.

If you continue to experience issues with the YOLOv5 model, the Ultralytics team offers support and guidance through our documentation at https://docs.ultralytics.com/yolov5. Good luck, and feel free to reach out with any questions!
